### ISSUE TITLE
Request is not returned to ReturnUrl after authentication success

### ISSUE THREAD

**Issue by brgrz** [auth0/auth0-aspnet-owin#54]
When requesting a protected resource the querystring is correctly appended by the "returnUrl" parameter but then after the authentication was successful the response always gets sent to the same (root) URL.

For instance, when using MVC areas when I request a protected resource within an "admin" area 
/admin/dashboard the request would be redirected to the login action on the account controller like this
http://localhost/admin/auth0/login?ReturnUrl=%2Fadmin%2Fdashboard

then when user authenticates and the authenticated request comes back from Auth0 the user is redirected to the http://localhost/ (not even to the /admin area). Should be redirected to the http://localhost/admin/dashboard

**Comment by jerriep** [comment]
@brgrz I have tested against our Quickstart sample, and this does indeed work correctly, and redirects the user to the protected resource after authentication.

Can you share the code you use in your `Login` method? You should ensure that pass the redirect URL to the `ChallengeResult`, e.g.

```
public ActionResult Login(string returnUrl)
{
    return new ChallengeResult("Auth0", returnUrl ?? Url.Action("Index", "Home"));
}
```

This is the code for `ChallengeResult`:

```
internal class ChallengeResult : HttpUnauthorizedResult
{
    private const string XsrfKey = "XsrfId";

    public ChallengeResult(string provider, string redirectUri)
        : this(provider, redirectUri, null)
    {
    }

    public ChallengeResult(string provider, string redirectUri, string userId)
    {
        LoginProvider = provider;
        RedirectUri = redirectUri;
        UserId = userId;
    }

    public string LoginProvider { get; set; }
    public string RedirectUri { get; set; }
    public string UserId { get; set; }

    public override void ExecuteResult(ControllerContext context)
    {
        var properties = new AuthenticationProperties { RedirectUri = RedirectUri };
        if (UserId != null)
        {
            properties.Dictionary[XsrfKey] = UserId;
        }
        context.HttpContext.GetOwinContext().Authentication.Challenge(properties, LoginProvider);
    }
}
```

Note that the `RedirectUri` should be added to the `AuthenticationProperties`

Also, are you using the latest version of the **Auth0-ASPNET-Owin** Nuget package?

**Comment by brgrz** [comment]
I actually use customized, self hosted version of embedded Lock and Auth0 OWIN 2.3.0 package. My GET Login method on the AccountController only returns the customized view - I do not use the ChallengeResult (I saw in a sample MVC app that ChallengeResult goes to Auth0 hosted login form, which I do not want). 

I'll try to post some gists for you to examine.

**Comment by jerriep** [comment]
The problem is that you are working outside of the standard authorization middleware, but initiating the authorization process from your own embedded Lock, and not initiating it through the ASP.NET middleware and the hosted Lock. So you will have to probably do some magic on your side to make it work.

I can try and assist you - as time allows - but I will need a full reproducible sample I can run on my ccomputer, please.

**Comment by brgrz** [comment]
I made a gist which is available here
https://gist.github.com/brgrz/c250d26fc6dd3d6aa2b6237c53111284

This is pretty much everything I got. I do have another project which also uses embedded Lock from last year and it uses the old (1.0.4) version of Auth0 Owin package. The same issue is present there, authentication works fine but redirection back to the corrent Url doesn't happen.

**Comment by brgrz** [comment]
> The problem is that you are working outside of the standard authorization middleware, but initiating the authorization process from your own embedded Lock, and not initiating it through the ASP.NET middleware and the hosted Lock. So you will have to probably do some magic on your side to make it work.

Yeah, I want to make the self hosted option to work. As far as I remember, there used to be samples for this even with ASP.NET/OWIN scenario but I couldn't find them anymore (I didn't come up with what I got by myself, I'm pretty sure it was published by Auth0 at some point but apparently removed later on).

**Comment by jerriep** [comment]
@brgrz please create a runnable sample application for me.

**Comment by jerriep** [comment]
Yeah, the embedded samples were removed as they were insecure. Unfortunately I do not have the exact details, but I do know Auth0 prefers you using the hosted version as it is more secure. For exact details on why, I suggest you contact Auth0 support (I am not an Auth0 employee)

**Comment by brgrz** [comment]
Oh, I wasn't aware of the fact you weren't an Auth0 employee, so thanks for helping out.

Given that everything changed in the last year or so since I used Auth0 for the first time and everyhint seems to point to using the hosted login now, I am going to evaluate if that would indeed be a better option going forward.

**Comment by jerriep** [comment]
I used to work for them, but not anymore. However, I am still maintaining these libraries for them, and they are paying me to do it - so this is my job. Well, one of them ;) So I will assist you where I can, but as I mentioned before

1. What you are doing is definitely not recommended by Auth0
2. What you are doing works outside the framework of the ASP.NET Authentication middleware, so it may not be easy to get it to do what you can

Bottom line: I would very much prefer you use the approach outlined in the ASP.NET (OWIN) Quickstarts and use the OWIN middleware it was intended to

**Comment by brgrz** [comment]
I understand but they do still mention it is possible to use the embedded Lock and I also see you have a sample for ASP.NET Core which uses self hosted login page (although not made with Lock but custom login form). I also have in another (working) project a secure code with the use of Html.AntiForgeryToken() and ValidateAntiforgeryTokenAttribute and the Owin set up to support it, it's just not straightforward. The only issue I'm running into is the returnUrl.

As said I will evaluate my options today whether the hosted login page is for us.

Thanks!

**Comment by jerriep** [comment]
Just a clarification on the ASP.NET Core sample you reference; That particular sample uses - as you mentioned - a custom MVC form. It then uses the Resource Owner Password grant to authenticate the user. 

So while that may appear on the surface to be a custom login similar to self-hosted Lock, it is not. The Resource Owner Password grant is completely supported by Auth0 and is very different than when you use self-hosted (or embedded) Lock

**Comment by jerriep** [comment]
@brgrz Any feedback on this

**Comment by brgrz** [comment]
I used my old code, tweaked it a bit and it works but only redirects back to root, so I had to change my app up a bit. Due to the way Auth0 segregates users across clients and connections (a bit akward for us) I am not sure whether we'll continue to use Auth0 in the long run. Add to that I'm waiting for a response from the Sales team for 3 days now, it's another reason why I'm not too thrilled about it.
